### PR TITLE
Hw9 persistency

### PR DESCRIPTION
**Pull request by alexey-chaichuk**
При запуске приложения считываем последний список фильмов из локальной БД. После этого запрашиваем список популярных фильмов с сайта. После его получения, заменяем данные в локальной БД на полученные из сети.
На скриншоте показан момент, когда на экране отображаются данные из локальной БД и идет запрос обновленных данных по сети.

Screenshot_1611301525

### CODE REVIEW

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesDataBase.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import android.content.Context
+import androidx.room.Database
+import androidx.room.Room
+import androidx.room.RoomDatabase
+import ru.chaichuk.hwapp.data.Movie
+
+@Database(entities = [Movie::class], version = 1)
+abstract class MoviesDataBase : RoomDatabase() {
+
+    abstract val moviesListDao : MoviesListDao
```
Comment (Axrorxoja) [inline app/src/main/java/ru/chaichuk/hwapp/db/MoviesDataBase.kt]: that's a good idea, then the abstract method

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    @TypeConverter
+    fun fromGenres(genres : List<Genre>) : String {
+        val gson = Gson()
+        val type = object : TypeToken<List<Genre>>() {}.type
+        return gson.toJson(genres, type)
+    }
+
+    @TypeConverter
+    fun toGenres(genresJson : String) : List<Genre> {
```
Comment (Axrorxoja) [inline app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt]: that's a bad idea
in the movie has only one genre 
and that you need to create a genre table and save genre id inside a movie

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        val gson = Gson()
+        val type = object : TypeToken<List<Genre>>() {}.type
+        return gson.fromJson(genresJson, type)
+    }
+
+
+    @TypeConverter
+    fun fromActors(actors : List<Actor>) : String {
+        val gson = Gson()
```
Comment (Axrorxoja) [inline app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt]: here also same moment

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesDataBase.kt (hunk 1 above)
Comment (alexey-chaichuk) [reply app/src/main/java/ru/chaichuk/hwapp/db/MoviesDataBase.kt]: I copied this code from workshop. And I saw it in other examples. Is there something wrong with it?

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt (hunk 2 above)
Comment (alexey-chaichuk) [reply app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt]: I did not find any solution to store one-to-many in room framework without adding new entity classes and additional classes with relations. So I decided to make less code and store all that data as text in the same record. 
In next commit I will add that additional tables and data classes

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesDataBase.kt (hunk 1 above)
Comment (Axrorxoja) [reply app/src/main/java/ru/chaichuk/hwapp/db/MoviesDataBase.kt]: no no, I just wont to say that good job :)

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt (hunk 2 above)
Comment (alexey-chaichuk) [reply app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt]: made new commit with separate tables for each entity

File: app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt (hunk 3 above)
Comment (alexey-chaichuk) [reply app/src/main/java/ru/chaichuk/hwapp/db/MoviesTypeConverters.kt]: made new commit with separate tables for each entity

File: app/src/main/java/ru/chaichuk/hwapp/viewModels/MoviesListViewModel.kt [hunk 4]
```diff
@@ -75,7 +75,7 @@ class MoviesListViewModel : ViewModel(){
 
                 moviesDbRepository.saveAllMovies(movies)
 
-                _mutableMoviesList.value = movies
+                _mutableMoviesList.value = movies.sortedBy { it.ratings }.asReversed()
```
Comment (Axrorxoja) [inline app/src/main/java/ru/chaichuk/hwapp/viewModels/MoviesListViewModel.kt:78]: better sort on SQL

Comment (alexey-chaichuk) [reply app/src/main/java/ru/chaichuk/hwapp/viewModels/MoviesListViewModel.kt:78]: That list is from network API and it is received unsorted. I sort list from DB in SQL, so I've added this sorting to make this two lists (from api and from db) with the same orders.

Comment (Axrorxoja) [reply app/src/main/java/ru/chaichuk/hwapp/viewModels/MoviesListViewModel.kt:78]: okay  I got you 
in this case all is ok

**Pull request merged**